Fix safe_constantize to not raise a LoadError.

kbrock opened the pull request:
### Summary

There was an issues when using `safe_constantize` on a string that has the wrong case.

File `em.rb` defines `EM`.
`"Em".safe_constantize` causes a little confusion with the autoloader.
The autoloader finds file "em.rb", expecting it to define `Em`, but `Em` is not defined.
The autoloader raises a `LoadError`, which is good, but `safe_constantize` is defined to return `nil` when a class is not found.

### Before

```ruby
"Em".safe_constantize
LoadError: Unable to autoload constant Em, expected rails/activesupport/test/autoloading_fixtures/em.rb to define it
```

### After

```ruby
"Em".safe_constantize
# => nil
```

kbrock committed:
Fix safe_constantize to not raise a LoadError.

### Summary

There was an issues when using `safe_constantize` on a string that has
the wrong case.

File `em.rb` defines `EM`.
`"Em".safe_constantize` causes a little confusion with the autoloader.
The autoloader finds file "em.rb",
expecting it to define `Em`, but `Em` is not defined.
The autoloader raises a `LoadError`, which is good,
But `safe_constantize` is defined to return `nil` when a class is not found.

### Before

```
"Em".safe_constantize
LoadError: Unable to autoload constant Em, \
expected rails/activesupport/test/autoloading_fixtures/em.rb to define it
```

### After

```
"Em".safe_constantize
# => nil
```

djberg96 commented on activesupport/lib/active_support/inflector/methods.rb:
```diff
@@ -328,6 +328,8 @@ def safe_constantize(camel_cased_word)
     rescue ArgumentError => e
       raise unless /not missing constant #{const_regexp(camel_cased_word)}!$/.match?(e.message)
+    rescue LoadError => e
+      rause unless /Unable to autoload constant/.match?(e.message)
```
rause?

gmcgibbon replied:
We might also want to add `#{const_regexp(camel_cased_word)}` to the end of this regex to ensure we're catching the right load error. Might be a good idea to test this actually raises now too seeing how CI is passing with a typo 🤔

gmcgibbon reviewed (commented):
I think this makes sense. 👍 Please add a changelog entry to ActiveSupport.

kbrock commented:
Thank you @gmcgibbon and @djberg96 

- added a test that made the `rause` fail (so it is properly tested)
- added `#{const_regexp(camel_cased_word)}` to regular expression to narrow down exception from the autoloader

gmcgibbon reviewed (commented):
Still needs a changelog entry [here](https://github.com/rails/rails/blob/master/activesupport/CHANGELOG.md). Other than that, LGTM! 👍

kbrock commented:
fixed Changelog conflict...

@gmcgibbon thank you so much for your help (and your patience for having to repeat that)

kbrock commented:
odd collate error for mysql (that I haven't seen before)
kicking

kbrock commented:
Thanks for all the help on this PR, and rails in general. It is hard work maintaining OSS.
